core/templates/pages/teach-page/teach-page.component.spec.ts

```diff
@@ -15,112 +15,197 @@
+        }
+        this._hashChange();
+      },
+      get href() {
+        return this._href;
+      },
+      set href(val) {
+        this._href = val;
+      },
+      reload: (val) => val
+    },
+    get onhashchange() {
+      return this.location._hashChange;
+    },
+
+    set onhashchange(val) {
+      this.location._hashChange = val;
+    }
+  };
+  get nativeWindow() {
+    return this._window;
+  }
+}
+
+class MockSiteAnalyticsService {
+  registerApplyToTeachWithOppiaEvent(): void {
+    return;
+  }
+}
+
+let component: TeachPageComponent;
+let fixture: ComponentFixture<TeachPageComponent>;
 
 describe('Teach Page', function() {
-  var $scope = null, ctrl = null;
-  var $timeout = null;
-  var SiteAnalyticsService = null;
-  var windowRef = new WindowRef();
-
-  beforeEach(angular.mock.module('oppia'));
-  beforeEach(angular.mock.module('oppia', function($provide) {
-    $provide.value('WindowRef', windowRef);
+  let windowRef: MockWindowRef;
+  let siteAnalyticsService = null;
+  let i18n = null;
+  let translate = null;
+
+  beforeEach(async(() => {
+    windowRef = new MockWindowRef();
+    TestBed.configureTestingModule({
+      declarations: [TeachPageComponent, MockTranslatePipe],
+      providers: [
+        {
+          provide: I18nLanguageCodeService,
+          useClass: MockI18nLanguageCodeService
+        },
+        { provide: SiteAnalyticsService, useClass: MockSiteAnalyticsService },
+        { provide: TranslateService, useClass: MockTranslateService },
+        UrlInterpolationService,
+        { provide: WindowRef, useValue: windowRef }
+      ],
+      schemas: [NO_ERRORS_SCHEMA]
+    }).compileComponents();
+    i18n = TestBed.get(I18nLanguageCodeService);
+    translate = TestBed.get(TranslateService);
+    siteAnalyticsService = TestBed.get(SiteAnalyticsService);
   }));
-  beforeEach(angular.mock.inject(function($injector, $componentController) {
-    $timeout = $injector.get('$timeout');
-    SiteAnalyticsService = $injector.get('SiteAnalyticsService');
-
-    var $rootScope = $injector.get('$rootScope');
-    $scope = $rootScope.$new();
 
-    ctrl = $componentController('teachPage', {
-      $rootScope: $scope
-    });
-  }));
+  beforeEach(() => {
+    fixture = TestBed.createComponent(TeachPageComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
 
-  afterEach(function() {
+  afterEach(() => {
     // Property onhashchange and location.hash are reassigned because it shares
     // same memory reference to all test blocks and the controller itself
     // because $provide.value of WindowRef refers to windowRef as well.
     // Once location.hash or onhashchange is setted in the controller,
     // the value will be only available in the test block itself, not affecting
     // others test block.
+    windowRef.nativeWindow.location.href = '';
     windowRef.nativeWindow.onhashchange = null;
     windowRef.nativeWindow.location.hash = '';
   });
 
```

Comment:
```suggestion
```

Remove all these lines, not required anymore.

Reply:
Done

Reply:
Why are we removing these?

Shouldn't we nullify all the location references before we proceed for each new check?

Reply:
Since we are using a mock service now, the window is not the global window variable, hence not shared between instances. So we don't need it according to me